pkg/webreg/webreg.go

```diff
@@ -0,0 +1,591 @@
+	</tr>
+</thead>
+<tbody>
+	{{ range .Steps }}
+		{{ tableEntryForStep . }}
+	{{ end }}
+</tbody>
+</table>
+<h2 id="source">TODO: Graphical representation that shows nested chain relationships</h2>
+`
+
+const WorkflowHTML = `
+<p><a href="/help">Back to index</a></p>
+<h1>Registry workflow "{{ .Name }}"</h1>
+<h2 id="documentation">Documentation</h2>
+{{ .Documentation }}
+<h2 id="pre" title="Steps run by this workflow to set up and configure the tests, in runtime order">Pre Steps</h2>
+<table class="table">
+<thead>
+	<tr>
+		<th title="The name of the reference or chain">Name</th>
+		<th title="The documentation for the reference or chain">Documentation</th>
+	</tr>
+</thead>
+<tbody>
+	{{ range .Pre }}
+		{{ tableEntryForStep . }}
+	{{ end }}
+</tbody>
+</table>
+<h2 id="test" title="Steps in the workflow that run actual tests, in runtime order">Test Steps</h2>
+<table class="table">
+<thead>
+	<tr>
+		<th title="The name of the reference or chain">Name</th>
+		<th title="The documentation for the reference or chain">Documentation</th>
+	</tr>
+</thead>
+<tbody>
+	{{ range .Test }}
+		{{ tableEntryForStep . }}
+	{{ end }}
+</tbody>
+</table>
+<h2 id="post" title="Steps run by this workflow to clean up and teardown test resources, in runtime order">Post Steps</h2>
+<table class="table">
+<thead>
+	<tr>
+		<th title="The name of the reference or chain">Name</th>
+		<th title="The documentation for the reference or chain">Documentation</th>
+	</tr>
+</thead>
+<tbody>
+	{{ range .Post }}
+		{{ tableEntryForStep . }}
+	{{ end }}
+</tbody>
+</table>
+<h2 id="source">TODO: Graph</h2>
+`
+
+const JobHTML = `
+<p><a href="/help">Back to index</a></p>
+<h1>Job "{{ .Name }}"</h1>
+<h2 id="pre" title="Steps run by this job to set up and configure the tests, in runtime order">Pre Steps</h2>
+<table class="table">
+<thead>
+	<tr>
+		<th title="The name of the reference or chain">Name</th>
+		<th title="The documentation for the reference or chain">Documentation</th>
+	</tr>
+</thead>
+<tbody>
+	{{ range .Pre }}
+		{{ tableEntryForStep . }}
+	{{ end }}
+</tbody>
+</table>
+<h2 id="test" title="Steps in the job that run actual tests, in runtime order">Test Steps</h2>
+<table class="table">
+<thead>
+	<tr>
+		<th title="The name of the reference or chain">Name</th>
+		<th title="The documentation for the reference or chain">Documentation</th>
+	</tr>
+</thead>
+<tbody>
+	{{ range .Test }}
+		{{ tableEntryForStep . }}
+	{{ end }}
+</tbody>
+</table>
+<h2 id="post" title="Steps run by this job to clean up and teardown test resources, in runtime order">Post Steps</h2>
+<table class="table">
+<thead>
+	<tr>
+		<th title="The name of the reference or chain">Name</th>
+		<th title="The documentation for the reference or chain">Documentation</th>
+	</tr>
+</thead>
+<tbody>
+	{{ range .Post }}
+		{{ tableEntryForStep . }}
+	{{ end }}
+</tbody>
+</table>
+<h2 id="source">TODO: Graph</h2>
+`
+
+type Components struct {
```

Comment:
I'm surprised we had to declare new types for this -- why couldn't the existing ones be used?

Reply:
Does that just move formatting/processing to Go code that creates these types instead of the template? I might prefer the latter if the logic is not too complex just to keep the types flat

Reply:
This was mainly done to reduce the number of functions I was using in the templates. This was more useful when I was testing. We could move more stuff into functions used by the templates to avoid new types.